### PR TITLE
Client capabilities

### LINKED ISSUE #174: [Feature Request] Client capabilities support
Smart Session Management (SSM) sometimes referred to as the long lived tokens (LLT) scenario require MSAL to allow developers to resolve challenges (formatted as claims request parameters) to be resolved silently rather than interactively.  Current releases of MSAL assume that interaction is required whenever a claims request parameter is included in the request.  

This change in behavior requires the client to indicate that it's aware of a new calling pattern relative to the use of MSAL.

See additional details of requirements in:
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1545

API/Design details:
https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=%2FClientCapabilities%2Foverview.md&version=GBdev
**Is your feature request related to a problem? Please describe.**

Related internal work item:
https://identitydivision.visualstudio.com/Engineering/_workitems/edit/824806

### PR DESCRIPTION
**Pull request by abhidnya13**
<sub>Linking back to relevant issue (on behalf of Abhi ^_^) for easier cross-check during reviewing</sub> This will close #174

### CODE REVIEW

File: tests/test_application.py [hunk 1]
```diff
@@ -267,3 +267,42 @@ def helper(scopes, account, authority, *args, **kwargs):
...
+        self.authority_url_in_app = "https://login.microsoftonline.com/common"
+        self.scopes = ["s1", "s2"]
+        self.client_id = "my_app"
+        self.claims = "{\"id_token\": {\"auth_time\": {\"essential\": true}}}"
+
+    def test_merged_claims_returned_correctly(self):
+        app = ClientApplication(
+            self.client_id,
+            authority=self.authority_url_in_app, client_capabilities=["llt", "ssm"])
+        merged_claims = "{\"id_token\": {\"auth_time\": {\"essential\": true}}, \"access_token\": {\"xms_cc\": {\"values\": [\"llt\", \"ssm\"]}}}"
```
[the remark this answers is not in the document]
Comment (Avery-Dunn) [reply tests/test_application.py:284]: My Python is rusty so I could be reading it wrong, but the format of the merged_claims value here looks correct, based on the values of client_capabilities and self.claims that I think get used in the test.

Check out the Claims/ClientCapabilities/ClientCapabilitiesAndClaimsJson values in [this constants file](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/master/tests/Microsoft.Identity.Test.Common/TestConstants.cs) of the .NET tests, or the very bottom of [this test](https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/dev/src/test/java/com/microsoft/aad/msal4j/AuthorizationRequestUrlParametersTest.java) in the Java tests for other examples of the claims/client capabilities input and the expected output of merging the two together.

File: msal/application.py [hunk 2]
```diff
@@ -249,6 +256,29 @@ def _build_client(self, client_credential, authority):
             on_removing_rt=self.token_cache.remove_rt,
             on_updating_rt=self.token_cache.update_rt)
 
+    def _merge_claims_and_capabilities(self, claims):
+        if claims:
+            claims = json.loads(claims)
```
[the remark this answers is not in the document]
Comment (abhidnya13) [reply msal/application.py:261]: All of the other SDKs take in claims as a json string. Line 299 should have been updated but regardless, we need to decide on what we want to take in as an input. I don't have a strong preference but if we want to be consistent across SDKs we can take it in as a string

File: tests/test_application.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (abhidnya13) [reply tests/test_application.py:284]: Yes I referred to these test cases as well, however both of these don not add  `xms_cc` part as a value in id_token field, I might be reading it wrong but I see in the above constants you mentioned we end up adding it only to `access_token` . However, as @rayluo  mentioned, in the [SSM](https://microsoft.sharepoint-df.com/:w:/t/SmartSessionManagementforOfficeClients/EU-SBJAwblNIlqlRwSFFDGUBbZng0DvLJAKNNB8Wy8hBKA?e=bKoiwM) doc mentioned it does say if you need both id_token and the access_token it needs to be added in both of these values. But from the client_capabilities [spec](https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=%2FClientCapabilities%2Foverview.md&version=GBdev&_a=preview) we only need to add it as an access_token claim so this probably should be clarified. But if you do follow client capabilities spec, it only relates xm_cc as access_token claim

File: msal/application.py [hunk 3]
```diff
@@ -79,6 +79,30 @@ def extract_certs(public_cert_content):
...
+            for key in client_capabilities_dict:
+                claims_node = claims.get(key, {})
+                if claims_node:
+                    claims_node.update(client_capabilities_dict[key])
+                else:
+                    claims[key] = client_capabilities_dict[key]
+
+        else:
+            return json.dumps(client_capabilities_dict)
+    return json.dumps(claims)
```
Comment (rayluo) [inline msal/application.py:103]: Currently, if claims input is `None`, the output would become `null`. This might not be what we want?

Also, we should have a test case for such edge case.

File: msal/application.py [hunk 4]
```diff
@@ -309,6 +346,7 @@ def get_authorization_request_url(
             self.http_client
             ) if authority else self.authority
 
+        claims = _merge_claims_and_capabilities(self.client_capabilities, claims)
```
Comment (rayluo) [inline msal/application.py:349]: Assigning a different value into the same input parameter is error-prone. I would suggest to either use a different variable name:

```suggestion
        _claims = _merge_claims_and_capabilities(self.client_capabilities, claims)
```

or see the comment below.

File: msal/application.py [hunk 5]
```diff
@@ -320,6 +358,7 @@ def get_authorization_request_url(
             scope=decorate_scope(scopes, self.client_id),
             nonce=nonce,
             domain_hint=domain_hint,
+            claims=claims,
```
Comment (rayluo) [inline msal/application.py:361]: (continue from above)
or we can do it inline:

```suggestion
            claims=_merge_claims_and_capabilities(self.client_capabilities, claims),
```

File: msal/application.py [hunk 6]
```diff
@@ -376,6 +421,7 @@ def acquire_token_by_authorization_code(
                 CLIENT_CURRENT_TELEMETRY: _build_current_telemetry_request_header(
                     self.ACQUIRE_TOKEN_BY_AUTHORIZATION_CODE_ID),
                 },
+            data=dict(kwargs.pop("data", {}), claims=claims),
```
Comment (rayluo) [inline msal/application.py:424]: Did we forget to merge claims beforehand?

File: msal/application.py [hunk 7]
```diff
@@ -830,16 +885,17 @@ def acquire_token_by_username_password(
...
             }
+        claims = _merge_claims_and_capabilities(self.client_capabilities, claims)
         if not self.authority.is_adfs:
             user_realm_result = self.authority.user_realm_discovery(
                 username, correlation_id=headers[CLIENT_REQUEST_ID])
             if user_realm_result.get("account_type") == "Federated":
                 return self._acquire_token_by_username_password_federated(
                     user_realm_result, username, password, scopes=scopes,
-                    headers=headers, **kwargs)
+                    headers=headers, data=dict(kwargs.pop("data", {}), claims=claims), **kwargs)
```
Comment (rayluo) [inline msal/application.py:895]: The new addition is already significantly long enough to occupy a line. We do not squeeze it into exist line, then. (Same suggestion on elsewhere, too.)
```suggestion
                    data=dict(kwargs.pop("data", {}), claims=claims),
                    headers=headers, **kwargs)
```

File: tests/test_application.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rayluo) [reply tests/test_application.py:277]: Documenting current decision: based on an offline mail thread with stakeholders, this topic is not yet fully settled. In such case, we will go with providing this Client Capability behavior for access_token only, because this gives us more flexibility when/if we will also introduce this to id_token at a later time.

File: tests/test_application.py [hunk 8]
```diff
@@ -267,3 +268,31 @@ def helper(scopes, account, authority, *args, **kwargs):
         with self.assertRaises(ExpectedBehavior):
             self.app.acquire_token_silent(["different scope"], self.account)
 
+
+class TestApplicationForClientCapabilities(unittest.TestCase):
+
+    def test_merged_claims_returned_correctly(self):
```
Comment (rayluo) [inline tests/test_application.py:274]: I believe our org encourages a more informational test method naming pattern "condition_a_b_c_should_result(...)" e.g. "test_divide_by_zero_should_raise_exception()". We might not always enforce that, but the current name is just too vague to differentiate it from its cohort. :-)
```suggestion
    def test_capability_and_id_token_claims(self):
```

File: tests/test_application.py [hunk 9]
```diff
@@ -267,3 +268,31 @@ def helper(scopes, account, authority, *args, **kwargs):
...
+        client_capabilities = ["llt", "ssm"]
+        claims = "{\"id_token\": {\"auth_time\": {\"essential\": true}}}"
```
Comment (rayluo) [inline tests/test_application.py:276]: Perhaps it is time to utilize Python's triple-quotation string? You can also use it to break down those long lines.
```suggestion
        claims = '''{"id_token": {"auth_time": {"essential": true}}}'''
```

File: msal/application.py [hunk 10]
```diff
@@ -79,6 +79,37 @@ def extract_certs(public_cert_content):
...
+    elif capabilities:
+        client_capabilities_dict = {
+            "access_token": {
+                "xms_cc": {
+                    "values": capabilities
+                }
+            }
+        }
+        return json.dumps(client_capabilities_dict)
+    return None
```
Comment (rayluo) [inline msal/application.py:110]: Utilize the full power of Python's dict to cut 30-ish lines into 6. :-)
```suggestion
def _merge_claims_and_capabilities(capabilities, claims):
    # Represent capabilities as {"access_token": {"xms_cc": {"values": capabilities}}}
    # and then merge/add it into incoming claims
    if capabilities:
        claims_dict = json.loads(claims) if claims else {}
        for key in ["access_token"]:  # We could add "id_token" if we'd decide to
            claims_dict.setdefault(key, {}).update(xms_cc={"values": capabilities})
        return json.dumps(claims_dict)
    return claims
```

File: msal/application.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rayluo) [reply msal/application.py]: Since we now decide to actually go with string, let's document so in each of the APIs.

File: msal/application.py [hunk 11]
```diff
@@ -795,7 +835,10 @@ def acquire_token_by_device_flow(self, flow, **kwargs):
...
         return self.client.obtain_token_by_device_flow(
             flow,
-            data=dict(kwargs.pop("data", {}), code=flow["device_code"]),
+            data=dict(
+                kwargs.pop("data", {}),
+                code=flow["device_code"],
+                claims=_merge_claims_and_capabilities(self._client_capabilities, claims)),
                 # 2018-10-4 Hack:
                 # during transition period,
                 # service seemingly need both device_code and code parameter.
```
Comment (rayluo) [inline msal/application.py:844]: My bad. Did not realize the previous change split this comment with its corresponding code
```suggestion
                code=flow["device_code"],  # 2018-10-4 Hack:
                    # during transition period,
                    # service seemingly need both device_code and code parameter.
                claims=_merge_claims_and_capabilities(self._client_capabilities, claims),
                ),
```

File: tests/test_application.py [hunk 12]
```diff
@@ -267,3 +268,41 @@ def helper(scopes, account, authority, *args, **kwargs):
...
+                        "access_token": {"nbf": {"essential": true, "value": "1563308371"},
+                                        "xms_cc": {"values": ["llt", "ssm"]}}}'''
+        # Comparing  dictionaries as JSON object order differs based on python version
+        self.assertEqual(
+            json.loads(merged_claims),
+            json.loads(_merge_claims_and_capabilities(client_capabilities, claims)))
+
+    def test_no_capabilities_only_claims_merge(self):
+        claims = '''{"id_token": {"auth_time": {"essential": true}}}'''
+        self.assertEqual(claims, _merge_claims_and_capabilities(None, claims))
```
Comment (rayluo) [inline tests/test_application.py:298]: The existing assertion is logically wrong. Let's be consistent to use your `json.loads(...)` pattern (which is great! Thanks!).
```suggestion
        self.assertEqual(
            json.loads(...),
            json.loads(_merge_claims_and_capabilities(...))
```

File: msal/application.py [hunk 13]
```diff
@@ -79,6 +79,16 @@ def extract_certs(public_cert_content):
...
 
+def _merge_claims_and_capabilities(capabilities, claims):
+    # Represent capabilities as {"access_token": {"xms_cc": {"values": capabilities}}}
+    # and then merge/add it into incoming claims
+    if capabilities:
+        claims_dict = json.loads(claims) if claims else {}
+        for key in ["access_token"]:  # We could add "id_token" if we'd decide to
+            claims_dict.setdefault(key, {}).update(xms_cc={"values": capabilities})
+        return json.dumps(claims_dict)
+    return claims
```
Comment (rayluo) [inline msal/application.py:90]: Less indentation is probably slightly better
```suggestion
    if not capabilities:
        return claims
    claims_dict = json.loads(claims) if claims else {}
    for key in ["access_token"]:  # We could add "id_token" if we'd decide to
        claims_dict.setdefault(key, {}).update(xms_cc={"values": capabilities})
    return json.dumps(claims_dict)
```

**Comment by abhidnya13** [comment]
As per internal discussion, changing the `claims` parameter to `claims_challenge` in the latest commit.

File: msal/application.py [hunk 14]
```diff
@@ -917,6 +1005,11 @@ def acquire_token_on_behalf_of(self, user_assertion, scopes, **kwargs):
 
         :param str user_assertion: The incoming token already received by this app
         :param list[str] scopes: Scopes required by downstream API (a resource).
+        :param claims_challenge:
+            The claims_challenge parameter requests specific claims requested by the resource provider
+            in the form of a claims_challenge directive in the www-authenticate header to be
+            returned from the UserInfo Endpoint and/or in the ID Token and/or Access Token.
+            It is a string of a JSON object which contains lists of claims being requested from these locations..
```
Comment (jiasli) [inline msal/application.py:1012]: Duplicate periods (.).

File: tests/test_application.py [hunk 15]
```diff
@@ -267,3 +268,45 @@ def helper(scopes, account, authority, *args, **kwargs):
         with self.assertRaises(ExpectedBehavior):
             self.app.acquire_token_silent(["different scope"], self.account)
 
+
+class TestApplicationForClientCapabilities(unittest.TestCase):
+
+    def test_capabilities_and_id_token_claims_merge(self):
+        client_capabilities = ["llt", "ssm"]
```
Comment (jiasli) [inline tests/test_application.py:275]: Are `llt` and `ssm` mocked-up values? Per the [doc](https://microsoft.sharepoint.com/:w:/t/aad/protocols/ERSWYtOQB45GgG4e2q3Cz00B6C36zi4gAs6JhdQG_wvVeQ?rtime=yEUj_nvd1kg&wdLOR=cB0CF5D1C-739A-4A91-983A-DBBDB9E775A6):

> The currently known values are: CP1.

This may confuse the readers.

File: msal/application.py (hunk 14 above)
Comment (rayluo) [reply msal/application.py:1012]: @abhidnya13 Jiashuo has sharp eyes... :-)

File: tests/test_application.py (hunk 15 above)
Comment (rayluo) [reply tests/test_application.py:275]: Good point. Perhaps we should have changed them into "foo", "bar". @abhidnya13

File: msal/application.py (hunk 14 above)
Comment (abhidnya13) [reply msal/application.py:1012]: He does :-) Jiashuo,  I have created a [new PR](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/250/files) addressing this and the other concern you have, please take a look :-) For some reason, I cannot add you as a reviewer.
